### PR TITLE
Reintroduce membership upgrade

### LINKED ISSUE #1744: Membership 'upgrade' to 'until graduation' is always 30 euros
### Describe the bug
Membership 'upgrade' to 'until graduation' is always 30 euros.

### How to reproduce
Steps to reproduce the behaviour:
1. Make sure you have a membership for the current year
2. Open the renewal page: http://localhost:8000/user/membership/
3. Select the member renewal with 'until graduation'
4. Submit
5. Open the admin and note that the contribution is 30 euros, it should give a discount since you already paid for this year

### Expected behaviour
The amount should include the discount and be 22,50.

### Screenshots
Screenshot 2021-06-20 at 19 05 57
Screenshot 2021-06-20 at 19 06 03

### PR DESCRIPTION
**Pull request by JobDoesburg**
Closes #1744 

### Summary
Current members, when upgrading their membership, will now not have to pay €30 but €22,5

### How to test
1. Have a member with a 1 year membership
2. Let them make a renewal
3. Profit

### CODE REVIEW

File: website/registrations/models.py [hunk 1]
```diff
@@ -111,7 +111,8 @@ def save(
         if self.membership_type == Membership.BENEFACTOR:
             self.length = self.MEMBERSHIP_YEAR
         else:
-            self.contribution = settings.MEMBERSHIP_PRICES[self.length]
+            if not self.contribution:
```
Comment (pingiun) [inline website/registrations/models.py:114]: This seems to cause the failing test, is it a necessary change?

File: website/registrations/models.py [hunk 2]
```diff
@@ -407,6 +408,19 @@ def save(
     ):
         if self.pk is None:
             self.status = Entry.STATUS_REVIEW
+
+        if (
+            self.contribution is None
+            and self.length == Entry.MEMBERSHIP_STUDY
+            and self.member.current_membership is not None
+            and self.member.current_membership.length - Entry.MEMBERSHIP_YEAR
```
Comment (pingiun) [inline website/registrations/models.py:416]: `self.member.current_membership.length - Entry.MEMBERSHIP_YEAR`

is not a boolean check? if this means `!= 0`, please write this explicitely

Comment (se-bastiaan) [reply website/registrations/models.py:416]: This was the previous check that we had: https://github.com/svthalia/concrexit/blob/ca6abb32946c961fa7d52bc04b1cd34bfb73e88e/website/registrations/services.py#L254


I believe `Entry.MEMBERSHIP_YEAR` is a string.

File: website/registrations/models.py [hunk 3]
```diff
@@ -111,7 +128,14 @@ def save(
         if self.membership_type == Membership.BENEFACTOR:
             self.length = self.MEMBERSHIP_YEAR
         else:
-            self.contribution = settings.MEMBERSHIP_PRICES[self.length]
+            if self.membership_upgrade_discount_applies:
+                # The membership upgrade discount applies
```
Comment (se-bastiaan) [inline website/registrations/models.py:132]:
```suggestion
```

Completely redundant

File: website/registrations/models.py [hunk 4]
```diff
@@ -111,7 +128,13 @@ def save(
         if self.membership_type == Membership.BENEFACTOR:
             self.length = self.MEMBERSHIP_YEAR
         else:
-            self.contribution = settings.MEMBERSHIP_PRICES[self.length]
+            if self.membership_upgrade_discount_applies:
+                self.contribution = (
+                    settings.MEMBERSHIP_PRICES["study"]
```
Comment (se-bastiaan) [inline website/registrations/models.py:133]: I think that these keys for the lengths are constants somewhere

Comment (JobDoesburg) [reply website/registrations/models.py:133]: I checked and we use the strings everywhere in the code base (as these are settings, not values in the database)

Comment (JobDoesburg) [reply website/registrations/models.py:133]: Will address in issue

**Review by pingiun — APPROVED** [review]
LGTM

**Pull request merged**